### PR TITLE
Fix the order of OUTPUT params in a reordered parameter list of procedures

### PR DESCRIPTION
**Pull request by basasairohan**
### Description

Currently, we are not handling the OUTPUT parameters in reordered parameter list of procedure arguments. This change adds support for handling the same. 

Signed-off-by: Sai Rohan Basa [bsrohan@amazon.com](mailto:bsrohan@amazon.com)

### Issues Resolved

BABEL-4235

### Test Scenarios Covered ###
* **Use case based -** Yes

* **Boundary conditions -** Yes

* **Arbitrary inputs -** Yes

* **Negative test cases -** N/A

* **Minor version upgrade tests -** N/A

* **Major version upgrade tests -** N/A

* **Performance tests -** Tested locally. It takes around 150ms for 100 arguments. Did not find much difference than the previous implementation

* **Tooling impact -** N/A

* **Client tests -** N/A

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/pl_exec-2.c [hunk 1]
```diff
@@ -877,15 +878,32 @@ exec_stmt_exec(PLtsql_execstate *estate, PLtsql_stmt_exec *stmt)
...
+				bool 	isFound = false;
 
 				if (argmodes &&
 					(argmodes[i] == PROARGMODE_INOUT ||
 					 argmodes[i] == PROARGMODE_OUT))
 				{
+					ListCell *paramcell;
+					j = 0;
+
+					foreach(paramcell, stmt->params)
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c:895]: Can we add a comment here?

File: contrib/babelfishpg_tsql/src/pl_exec-2.c [hunk 2]
```diff
@@ -877,15 +878,32 @@ exec_stmt_exec(PLtsql_execstate *estate, PLtsql_stmt_exec *stmt)
...
+					{
+						tsql_exec_param *p = (tsql_exec_param *) lfirst(paramcell);
+						if (argnames[i] && p->name && strcmp(argnames[i], p->name) == 0)
+						{
+							isFound = true;
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: Do we really need another variable for this?
Why can't you initialise j to -1 and then if you match this condition then set j.
Use j conditionally later

File: contrib/babelfishpg_tsql/src/pl_exec-2.c [hunk 3]
```diff
@@ -877,15 +878,32 @@ exec_stmt_exec(PLtsql_execstate *estate, PLtsql_stmt_exec *stmt)
...
+							break;
+						}
+						j++;
+					}
+
+					if (!isFound) j = i;
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: As a thumb rule in Postgres and babelfish we follow multi line if/else statements

File: contrib/babelfishpg_tsql/src/pl_exec-2.c [hunk 4]
```diff
@@ -815,7 +815,8 @@ exec_stmt_exec(PLtsql_execstate *estate, PLtsql_stmt_exec *stmt)
 			PLtsql_row *row;
 			int			nfields;
 			int			i;
-			ListCell   *lc;
+			int			j;
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/src/pl_exec-2.c]: Rename j to a more logical variable name - eg relativeArgIndex

File: test/JDBC/expected/TestStoredProcedures.out [hunk 5]
```diff
@@ -239,3 +239,152 @@ DROP PROCEDURE sp_test14
 #storedproc#!#prep#!#sp_test19#!#uniqueidentifier|-|a|-|5b7c2e8d-6d90-411d-8e19-9a81067e6f6c|-|output
 #storedproc#!#prep#!#sp_test19#!#uniqueidentifier|-|a|-|5b7c2e8d-6d90-411d-8e19-9a81067e6f6c|-|inputoutput
 #DROP PROCEDURE sp_test19
+
+CREATE PROCEDURE sp_test20 (@a INT, @b INT OUTPUT) AS BEGIN Select @a as a, @b as b; SET @b=100; SET @a=1000; END;
```
Comment (KushaalShroff) [inline test/JDBC/expected/TestStoredProcedures.out]: Please add RPC tests as well, Just take the same sql batch and write the prep-exec cases.

Comment (KushaalShroff) [inline test/JDBC/expected/TestStoredProcedures.out]: Since this code change is touching all the param modes, can we add tests for those as well?

**Review by KushaalShroff — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/pl_exec-2.c (hunk 1 above)
Comment (basasairohan) [reply contrib/babelfishpg_tsql/src/pl_exec-2.c:895]: Added.

**Review by KushaalShroff — APPROVED** [review]

**Pull request merged**